### PR TITLE
candidate totals by party by office

### LINKED ISSUE #3549: Make new /candidates/totals/by_party/by_office endpoint
**What we're after:** In order to support the party aggregate totals work, we need to make new 
/candidates/totals/by_party/by_office endpoint. 

**Completion criteria:**
- [x] Identify table that houses the data for receipts and disbursements
- [x] Create endpoint that can provide both receipts and disbursement information
- [x] Determine what data points need to be filterable
- [x] Include additional filter for is_active_candidate
- [x] Compare and make sure the other totals match up

Design work happening here: https://github.com/fecgov/fec-cms/issues/2620

### PR DESCRIPTION
**Pull request by fecjjeng**
## Summary (required)

- Resolves #3549 Make new /candidates/totals/by_party/by_office endpoint

Data Source:
ofec_candidate_totals_mv has the financial information aggregated by candidate/election_cycles available already. It does not have the office, party, and candidate_active information though.
However, ofec_candidate_history_with_future_election_vw is the driver in the main query,
which already has all the additional candidate information needed:
office
party
candidate_active
Add these 3 fields in the list of select.

new endpoint named /candidates/totals/by_office/by_party/ had been created.

## How to test the changes locally
A tmp version of the updated mv (ofec_candidate_totals_mv_tmp_jj) had been created in DEV database. 

Connected to the DEV database and execute the following query, it should show the total receipts and total disbursement for all active candidates of election_year 2020 for office= 'P': 
SELECT ofec_candidate_totals_mv_tmp_jj.office AS office, 
CASE WHEN (ofec_candidate_totals_mv_tmp_jj.party = 'DFL') THEN 'DEM' WHEN (ofec_candidate_totals_mv_tmp_jj.party = 'DEM') THEN 'DEM' WHEN (ofec_candidate_totals_mv_tmp_jj.party = 'REP') THEN 'REP' ELSE 'Other' END AS party,
 ofec_candidate_totals_mv_tmp_jj.election_year AS election_year, 
 sum(ofec_candidate_totals_mv_tmp_jj.receipts) AS total_receipts, 
 sum(ofec_candidate_totals_mv_tmp_jj.disbursements) AS total_disbursements 
FROM ofec_candidate_totals_mv_tmp_jj 
WHERE ofec_candidate_totals_mv_tmp_jj.is_election = true AND ofec_candidate_totals_mv_tmp_jj.office = 'P' AND ofec_candidate_totals_mv_tmp_jj.election_year IN (2020) 
AND ofec_candidate_totals_mv_tmp_jj.candidate_inactive = False
GROUP BY ofec_candidate_totals_mv_tmp_jj.office, 
CASE WHEN (ofec_candidate_totals_mv_tmp_jj.party = 'DFL') THEN 'DEM' WHEN (ofec_candidate_totals_mv_tmp_jj.party = 'DEM') THEN 'DEM' WHEN (ofec_candidate_totals_mv_tmp_jj.party = 'REP') THEN 'REP' ELSE 'Other' END, ofec_candidate_totals_mv_tmp_jj.election_year; 

Download this branch to local machine. 
In the model file candidates.py, change the following line to point to the tmp mv
class CandidateTotal(db.Model):
    __tablename__ = 'ofec_candidate_totals_mv_tmp_jj'

Start the local server. 
Under candidate, looking for the new endpoint
/candidates/totals/by_office/by_party/ 

Tested with election_year/office/party.  
2020/P/true (the result should match the query above)
2020/S/true (you can modify the query above to verify the result)
or other combinations you prefer.

Also check is_active_candidate
if no value specified, all candidates data is returned (in the query, AND ofec_candidate_totals_mv_tmp_jj.candidate_inactive = True/False is omitted). 
When True is specified, only active candidates data are returned (in the query, candidate_inactive = False). 
When False is specified, only inactive candidates data is returned (in the query, candidate_inactive = True).

## Impacted areas of the application
List general components of the application that this PR will affect:

New endpoint under "candidate"
/candidates/totals/by_office/by_party/ 

## Related PRs
List related PRs against other branches:

branch | PR
------ | ------
fix/other_pr | [link]()
feature/other_pr | [link]()

### CODE REVIEW

**Comment by hcaofec** [comment]
I ran this query 
**select distinct party from ofec_candidate_totals_mv_tmp_jj**
and it returns 148 rows including a NULL value.  Should the output of this endpoint consolidate the party type into  DEM, REP, and OTHERS like what front end has, or leave the calculation to front end? 
@fec-jli @lbeaufort @fecjjeng 


Screen Shot 2019-04-11 at 7 13 29 PM

**Comment by fecjjeng** [comment]
@hcaofec Good question.  I did not cross check the cms ticket.  Did not notice the need for Other and All.  If the sole purpose of this endpoint is to feed that front page, then my output need to be changed to DEM/REP/Other, which I have an update that is cable to do so I can push in.  
Then it occurs to me, if that is the case, maybe we should NOT let the user filter on "party"?   If we do, then In case users does put in a party as a filter, the output will be confusing.

**Comment by jwchumley** [comment]
What front-end feature are you trying to support? If you do use dem, rep and others there is a gotcha. DFL should appear as a DEM in this case. There may be others - Paul would know or you could look at the sql for the classic map.


Jeff Chumley

Electronic Filing, Web and Disclosure Branch Manager

Federal Election Commission



________________________________

**Comment by fecjjeng** [comment]
If it is just on the output, I can easily calculate DFL as DEM.  Actually, I have an update ready to do so that I can push.

My major concern is that if we allow user to filter on “party”, then

If they input “DEM”, there will only be one row DEM appears.
If they put in “GRE”, there will be only one row as “Other” appear.
If they put in “Other” or “DFL", there will be no matching data for it.

So if the sole purpose is of this endpoint is to feed that front page, then my output need to be changed to DEM/REP/Other.  And not allow “party” as a filter.  Or does not add this end point to rest.py so user will not see it.


On Apr 12, 2019, at 7:50 AM, Jeff Chumley <notifications@github.com<mailto:notifications@github.com>> wrote:

What front-end feature are you trying to support? If you do use dem, rep and others there is a gotcha. DFL should appear as a DEM in this case. There may be others - Paul would know or you could look at the sql for the classic map.


Jeff Chumley

Electronic Filing, Web and Disclosure Branch Manager

Federal Election Commission



________________________________

**Comment by PaulClark2** [comment]
@jwchumley we are trying to support a "horse race" feature. @hcaofec @fecjjeng it makes sense to me that we should be able to display "all" as well as the three party breakdowns: DEM, REP and OTHER.  Democratic Party includes both DEM and DFL, Republican Party only includes REP and Other is all parties not DEM, DFL or REP.  

If we are providing the ability to click through these high-level aggregates, when a user clicks on other will we be able to display all except DEM, DFL and REP? I think we have an exclude function that allows that on other endpoints.

**Comment by lbeaufort** [comment]
@PaulClark2 I'd want to check with @JonellaCulmer to see if we intend to have click-through capabilities on the party-level aggregates. We can't currently filter schedule A by candidate party, and at first thought it would be pretty extensive to do that. The current design looks like just a visual, not a click-through: https://github.com/fecgov/fec-cms/issues/2620

**Comment by PaulClark2** [comment]
@lbeaufort these would be report level totals instead of schedule-level totals, right? Or, am I misunderstanding how these office/party totals are being generated.

**Comment by lbeaufort** [comment]
After our discussion in design sync, we don't need to click through to schedule data. If we ever wanted to do a click-through, we'd probably want to click through to the candidate totals data tables (/candidates/totals)

**Comment by fecjjeng** [comment]
After today's design meeting, it is clarified that this new endpoint will mainly support the aggregate in frontend page (fecgov/fec-cms#2620). It is not needed to let the user filter by party.  The output will be grouped into DEM/REP/Other per election_year.  Note:  DFL will be grouped together as DEM.

File: webservices/common/models/candidates.py [hunk 1]
```diff
@@ -143,7 +143,9 @@ class CandidateTotal(db.Model):
     coverage_end_date = db.Column(db.Date, doc=docs.COVERAGE_END_DATE)
     federal_funds_flag = db.Column(db.Boolean, index=True, doc=docs.FEDERAL_FUNDS_FLAG)
     has_raised_funds = db.Column(db.Boolean, index=True, doc=docs.HAS_RAISED_FUNDS)
-
+    party = db.Column(db.String(3), index=True, doc=docs.PARTY)
+    office = db.Column(db.String(1), index=True, doc=docs.OFFICE)
+    candidate_inactive = db.Column(db.Boolean, doc='True indicates that a candidate is inactive.')
```
[the remark this answers is not in the document]
Comment (fecjjeng) [reply webservices/common/models/candidates.py]: Done

File: webservices/resources/candidate_aggregates.py [hunk 2]
```diff
@@ -261,3 +261,74 @@ def build_query(self, **kwargs):
...
+        )
+
+        if kwargs.get('office') and kwargs['office'] is not None:
+            query = query.filter(
+                total.office == kwargs['office']
+            )
+        if kwargs.get('party') and kwargs['party'] is not None:
+            query = query.filter(
+                total.party == kwargs['party']
+            )
```
[the remark this answers is not in the document]
Comment (fecjjeng) [reply webservices/resources/candidate_aggregates.py]: Thank you for catch it.  Missed this one.
Updated the file and pushed again.

**Comment by fec-jli** [comment]
good work.
question, Are `election_year` and `office` mandatory input fields? if they are, we should handle these.

**Pull request merged**